### ISSUE TITLE
DurationBarAggregator fails with half backed bars

### ISSUE THREAD

**Issue by francescopeloi** [ta4j/ta4j#1243]
**I have checked existing issues and wiki**
- [x] I could not find similar [issues](https://github.com/ta4j/ta4j/issues?utf8=%E2%9C%93&q=)
- [x] I could not find a solution in the [wiki](https://ta4j.github.io/ta4j-wiki/) or [faq section](https://ta4j.github.io/ta4j-wiki/FAQ.html)

While backtesting a strategy that was using only the bar close price, to optimize memory I was creating bars with only the close price. The DurationBarAggregator fails as it tries to get a numFactory from the openPrice, that was not existing:
https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/aggregator/DurationBarAggregator.java#L94

Something like this would prevent the failure and still aggregate the bars with the prices that are found:
```
NumFactory getNumFactory(Bar bar) {
        return Stream.of(
                        bar.getClosePrice(),
                        bar.getOpenPrice(),
                        bar.getHighPrice(),
                        bar.getLowPrice(),
                        bar.getVolume(),
                        bar.getAmount()
                )
                .filter(Objects::nonNull)
                .map(Num::getNumFactory)
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("No valid NumFactory found in the bar"));
}
```
This makes it work also with "half-backed" bars. If it feels reasonable I can implement.

**Comment by francescopeloi** [comment]
or is there a better way to get the NumFactory? I guess it's taking it out of a bar to make sure it's getting the _right_ NumFactory? (DoubleNum or DecimalNum)

**Comment by sgflt** [comment]
I did a small workaround shaped as NumFactoryProvider. But it has limitation that restricts all simulations to be run with the same Factory instance.

So it is not possible to run one simulation with precision of 10 and another with precison of 20 in the same JVM.

https://github.com/sgflt/ta4j/blob/96e6d5405a35464872ace78b527456272541a9c3/ta4j-core/src/main/java/org/ta4j/core/num/NumFactoryProvider.java#L10-L15

**Comment by francescopeloi** [comment]
Thanks for your comment @sgflt, but I am not sure how it'd help: I could do `DoubleNumFactory.getInstance()` already, without going through another class, no? But it doesn't assure that it'd be using the same numFactory used to create the bar prices, and in my understanding that is the problem.

**Comment by sgflt** [comment]
NumFactoryProvider is central point where you can set desired NumFactory before any execution. Then all instances of num are created by this predefined factory.

```java
NumFactoryProvider.setDefaultNumFactory(DecimalNumFactory.getInstance())
```

https://github.com/sgflt/ta4j/blob/96e6d5405a35464872ace78b527456272541a9c3/ta4j-core/src/main/java/org/ta4j/core/api/Indicators.java#L61-L63

I am thinking about better mechanism, that allows to define precision per simulation context. But if you don't need change precisions, it will be sufficient.

**Comment by sgflt** [comment]
Also I wonder what happens if two instances with different precisions are mixed together. Is preserved the higher precision? If yes then all this passing of NumFactories may be cut off.

**Comment by sgflt** [comment]
```java
import java.math.BigDecimal;
import java.math.MathContext;

class Scratch {
  public static void main(String[] args) {
    final var doubleDecimal = BigDecimal.valueOf(0.1);
    var bigDecimal =
        new BigDecimal("0.100000000000000000000000000000000000000000000000000000000000000000", new MathContext(128));

    System.out.println("double source: " + doubleDecimal);
    System.out.println("string source: " + bigDecimal);
    System.out.println("string precision: " + bigDecimal.precision());
    System.out.println("double precision: " + doubleDecimal.precision());
    final var add = doubleDecimal.add(bigDecimal);
    System.out.println("sum: " + add);
    System.out.println("sum precision: " + add.precision());
  }

}

```

```
double source: 0.1
string source: 0.100000000000000000000000000000000000000000000000000000000000000000
string precision: 66
double precision: 1
sum: 0.200000000000000000000000000000000000000000000000000000000000000000
sum precision: 66
```


Why we pass NumFactories further down? I have refactored Num creation with Cargo pattern... now I think it may be simplified much more. I don't know from where I though there may be some precision loss. 

The opposite is true, precision may be unnecessaryly too big, so we need factory that sets upper limit for precision.

In your case you need only constant zero, so it may be possible just throw away the caching map with various precisions and hardcode zero constant 0 with precision 0. For choosing the proper Num backend may be used mentioned NumFactoryProvider.

**Comment by sgflt** [comment]
https://github.com/sgflt/ta4j/commit/c37fcbc620cf5818bf570623c08a253c7048673b

Is it correct?

**Comment by francescopeloi** [comment]
Of course, I am doing this because I am sure the open price is not used on the indicators I am currently using. This is some non-indicator logic that is breaking though.

**Comment by francescopeloi** [comment]
> [sgflt@6c1cc2b](https://github.com/sgflt/ta4j/commit/6c1cc2b3f3ce7423cd28807c4bc8a11695d7d79f)
> 
> Is it correct?

sorry if I haven't answered you @sgflt , but it takes me some time to understand the code you created and I haven't found the time yet.
